// @blocknote/xl-multi-column/dist/blocknote-xl-multi-column.js
import { BlockNoteSchema as e, UniqueID as t, createBlockSpecFromTiptapNode as n, createExtension as r, getBlockInfo as i, getNearestBlockPos as a, getNodeById as o, nodeToBlock as s } from "@blocknote/core";
import { Plugin as c, PluginKey as l } from "prosemirror-state";
import { Extension as u, Node as d } from "@tiptap/core";
import { SideMenuExtension as f, insertOrUpdateBlockForSlashMenu as p } from "@blocknote/core/extensions";
import { Decoration as m, DecorationSet as h } from "prosemirror-view";
import { TbColumns2 as g, TbColumns3 as _ } from "react-icons/tb";
import { jsx as v } from "react/jsx-runtime";
//#region \0rolldown/runtime.js
var y = Object.defineProperty, b = (e, t) => {
	let n = {};
	for (var r in e) y(n, r, {
		get: e[r],
		enumerable: !0
	});
	return t || y(n, Symbol.toStringTag, { value: "Module" }), n;
}, x = { slash_menu: {
	two_columns: {
		title: "عمودان",
		subtext: "عمودان جنبًا إلى جنب",
		aliases: [
			"أعمدة",
			"صف",
			"تقسيم"
		],
		group: "الكتل الأساسية"
	},
	three_columns: {
		title: "ثلاثة أعمدة",
		subtext: "ثلاثة أعمدة جنبًا إلى جنب",
		aliases: [
			"أعمدة",
			"صف",
			"تقسيم"
		],
		group: "الكتل الأساسية"
	}
} }, S = { slash_menu: {
	two_columns: {
		title: "Zwei Spalten",
		subtext: "Zwei Spalten nebeneinander",
		aliases: [
			"Spalten",
			"Reihe",
			"teilen"
		],
		group: "Grundlegende Blöcke"
	},
	three_columns: {
		title: "Drei Spalten",
		subtext: "Drei Spalten nebeneinander",
		aliases: [
			"Spalten",
			"Reihe",
			"teilen"
		],
		group: "Grundlegende Blöcke"
	}
} }, C = { slash_menu: {
	two_columns: {
		title: "Two Columns",
		subtext: "Two columns side by side",
		aliases: [
			"columns",
			"row",
			"split"
		],
		group: "Basic blocks"
	},
	three_columns: {
		title: "Three Columns",
		subtext: "Three columns side by side",
		aliases: [
			"columns",
			"row",
			"split"
		],
		group: "Basic blocks"
	}
} }, w = { slash_menu: {
	two_columns: {
		title: "Dos Columnas",
		subtext: "Dos columnas lado a lado",
		aliases: [
			"columnas",
			"fila",
			"dividir"
		],
		group: "Bloques básicos"
	},
	three_columns: {
		title: "Tres Columnas",
		subtext: "Tres columnas lado a lado",
		aliases: [
			"columnas",
			"fila",
			"dividir"
		],
		group: "Bloques básicos"
	}
} }, T = { slash_menu: {
	two_columns: {
		title: "Deux Colonnes",
		subtext: "Deux colonnes côte à côte",
		aliases: [
			"colonnes",
			"rangée",
			"partager"
		],
		group: "Blocs de base"
	},
	three_columns: {
		title: "Trois Colonnes",
		subtext: "Trois colonnes côte à côte",
		aliases: [
			"colonnes",
			"rangée",
			"partager"
		],
		group: "Blocs de base"
	}
} }, E = { slash_menu: {
	two_columns: {
		title: "Dva Stupca",
		subtext: "Dva stupca jedan pored drugog",
		aliases: [
			"stupci",
			"redak",
			"podijeli"
		],
		group: "Osnovni blokovi"
	},
	three_columns: {
		title: "Tri Stupca",
		subtext: "Tri stupca jedan pored drugog",
		aliases: [
			"stupci",
			"redak",
			"podijeli"
		],
		group: "Osnovni blokovi"
	}
} }, D = { slash_menu: {
	two_columns: {
		title: "Tvær Dálkar",
		subtext: "Tvær dálkar hlið við hlið",
		aliases: [
			"dálkar",
			"röð",
			"skipta"
		],
		group: "Grunnblokkar"
	},
	three_columns: {
		title: "Þrír Dálkar",
		subtext: "Þrír dálkar hlið við hlið",
		aliases: [
			"dálkar",
			"röð",
			"skipta"
		],
		group: "Grunnblokkar"
	}
} }, O = { slash_menu: {
	two_columns: {
		title: "二列",
		subtext: "二列並んで",
		aliases: [
			"列",
			"行",
			"分割"
		],
		group: "基本ブロック"
	},
	three_columns: {
		title: "三列",
		subtext: "三列並んで",
		aliases: [
			"列",
			"行",
			"分割"
		],
		group: "基本ブロック"
	}
} }, k = { slash_menu: {
	two_columns: {
		title: "두 열",
		subtext: "두 열 나란히",
		aliases: [
			"열",
			"행",
			"분할"
		],
		group: "기본 블록"
	},
	three_columns: {
		title: "세 열",
		subtext: "세 열 나란히",
		aliases: [
			"열",
			"행",
			"분할"
		],
		group: "기본 블록"
	}
} }, A = { slash_menu: {
	two_columns: {
		title: "Twee Kolommen",
		subtext: "Twee kolommen naast elkaar",
		aliases: [
			"kolommen",
			"rij",
			"verdelen"
		],
		group: "Basisblokken"
	},
	three_columns: {
		title: "Drie Kolommen",
		subtext: "Drie kolommen naast elkaar",
		aliases: [
			"kolommen",
			"rij",
			"verdelen"
		],
		group: "Basisblokken"
	}
} }, j = { slash_menu: {
	two_columns: {
		title: "Dwie Kolumny",
		subtext: "Dwie kolumny obok siebie",
		aliases: [
			"kolumny",
			"rząd",
			"podzielić"
		],
		group: "Podstawowe bloki"
	},
	three_columns: {
		title: "Trzy Kolumny",
		subtext: "Trzy kolumny obok siebie",
		aliases: [
			"kolumny",
			"rząd",
			"podzielić"
		],
		group: "Podstawowe bloki"
	}
} }, M = { slash_menu: {
	two_columns: {
		title: "Duas Colunas",
		subtext: "Duas colunas lado a lado",
		aliases: [
			"colunas",
			"linha",
			"dividir"
		],
		group: "Blocos básicos"
	},
	three_columns: {
		title: "Três Colunas",
		subtext: "Três colunas lado a lado",
		aliases: [
			"colunas",
			"linha",
			"dividir"
		],
		group: "Blocos básicos"
	}
} }, N = { slash_menu: {
	two_columns: {
		title: "Два Столбца",
		subtext: "Два столбца рядом",
		aliases: [
			"столбцы",
			"ряд",
			"разделить"
		],
		group: "Базовые блоки"
	},
	three_columns: {
		title: "Три Столбца",
		subtext: "Три столбца рядом",
		aliases: [
			"столбцы",
			"ряд",
			"разделить"
		],
		group: "Базовые блоки"
	}
} }, P = { slash_menu: {
	two_columns: {
		title: "Hai Cột",
		subtext: "Hai cột cạnh nhau",
		aliases: [
			"cột",
			"hàng",
			"chia"
		],
		group: "Khối cơ bản"
	},
	three_columns: {
		title: "Ba Cột",
		subtext: "Ba cột cạnh nhau",
		aliases: [
			"cột",
			"hàng",
			"chia"
		],
		group: "Khối cơ bản"
	}
} }, F = { slash_menu: {
	two_columns: {
		title: "两列",
		subtext: "两列并排",
		aliases: [
			"列",
			"行",
			"分割"
		],
		group: "基础"
	},
	three_columns: {
		title: "三列",
		subtext: "三列并排",
		aliases: [
			"列",
			"行",
			"分割"
		],
		group: "基础"
	}
} }, I = /* @__PURE__ */ b({
	ar: () => x,
	de: () => S,
	en: () => C,
	es: () => w,
	fr: () => T,
	hr: () => E,
	is: () => D,
	ja: () => O,
	ko: () => k,
	nl: () => A,
	pl: () => j,
	pt: () => M,
	ru: () => N,
	vi: () => P,
	zh: () => F
});
//#endregion
//#region src/i18n/dictionary.ts
function L(e) {
	if (!e.dictionary.multi_column) throw Error("Multi-column dictionary not found");
	return e.dictionary.multi_column;
}
//#endregion
//#region src/extensions/DropCursor/multiColumnDropCursor.ts
var R = .1;
function z(e, t, n) {
	let r = t.posAtCoords({
		left: e.clientX,
		top: e.clientY
	});
	if (!r) return null;
	let i = a(n.doc, r.pos), o = n.doc.resolve(i.posBeforeNode);
	o.parent.type.name === "column" && (o = n.doc.resolve(o.before()));
	let s = {
		posBeforeNode: o.pos,
		node: o.nodeAfter
	}, c = t.nodeDOM(s.posBeforeNode);
	if (c === null) return {
		position: "regular",
		posBeforeNode: s.posBeforeNode,
		node: s.node
	};
	let l = c.getBoundingClientRect(), u = "regular";
	return e.clientX <= l.left + l.width * R ? u = "left" : e.clientX >= l.right - l.width * R && (u = "right"), {
		position: u,
		posBeforeNode: s.posBeforeNode,
		node: s.node
	};
}
var B = { hooks: { computeDropPosition: (e) => {
	let t = z(e.event, e.view, e.view.state);
	return t === null || t.position === "regular" ? e.defaultPosition : {
		pos: t.posBeforeNode,
		orientation: t.position === "left" ? "block-vertical-left" : "block-vertical-right"
	};
} } };
//#endregion
//#region src/extensions/DropCursor/multiColumnHandleDropPlugin.ts
function V(e) {
	return new c({ props: { handleDrop(n, r, a, o) {
		let c = z(r, n, n.state);
		if (c === null) return !1;
		let l = i(c);
		if (c.position === "regular" || a.content.childCount === 0) return !1;
		let u = s(a.content.child(0), e.pmSchema);
		if (l.blockNoteType === "column") {
			let r = s(n.state.doc.resolve(l.bnBlock.beforePos).node(), e.pmSchema), i = 0;
			r.children.forEach((e) => {
				i += e.props.width;
			});
			let a = i / r.children.length;
			if (a < .99 || a > 1.01) {
				let e = 1 / a;
				r.children.forEach((t) => {
					t.props.width = t.props.width * e;
				});
			}
			let o = r.children.findIndex((e) => e.id === l.bnBlock.node.attrs.id), d = r.children.map((e) => ({
				...e,
				children: e.children.filter((e) => e.id !== u.id)
			})).filter((e) => e.children.length > 0).toSpliced(c.position === "left" ? o : o + 1, 0, {
				type: "column",
				children: [u],
				props: {},
				content: void 0,
				id: t.options.generateID()
			});
			e.getBlock(u.id) && e.removeBlocks([u]), e.updateBlock(r, { children: d });
		} else {
			let t = s(l.bnBlock.node, e.pmSchema);
			if (t.id === u.id) return !0;
			let n = c.position === "left" ? [u, t] : [t, u];
			e.getBlock(u.id) && e.removeBlocks([u]), e.replaceBlocks([t], [{
				type: "columnList",
				children: n.map((e) => ({
					type: "column",
					children: [e]
				}))
			}]);
		}
		return !0;
	} } });
}
var H = r(({ editor: e }) => ({
	key: "multiColumnDropHandler",
	prosemirrorPlugins: [V(e)]
})), U = new l("ColumnResizePlugin"), W = class {
	editor;
	view;
	RESIZE_MARGIN_WIDTH_PX = 20;
	COLUMN_MIN_WIDTH_PERCENT = .5;
	constructor(e, t) {
		this.editor = e, this.view = t, this.view.dom.addEventListener("mousedown", this.mouseDownHandler), document.body.addEventListener("mousemove", this.mouseMoveHandler), document.body.addEventListener("mouseup", this.mouseUpHandler);
	}
	getColumnHoverOrDefaultState = (e) => {
		if (!this.editor.isEditable) return { type: "default" };
		let t = e.target;
		if (!this.view.dom.contains(t)) return { type: "default" };
		let n = t.closest(".bn-block-column");
		if (!n) return { type: "default" };
		let r = e.clientX, i = n.getBoundingClientRect(), a = r < i.left + this.RESIZE_MARGIN_WIDTH_PX ? "left" : r > i.right - this.RESIZE_MARGIN_WIDTH_PX ? "right" : "none", s = a === "left" ? n.previousElementSibling : a === "right" ? n.nextElementSibling : void 0;
		if (!s) return { type: "default" };
		let c = a === "left" ? s : n, l = a === "left" ? n : s, u = c.getAttribute("data-id"), d = l.getAttribute("data-id"), f = o(u, this.view.state.doc), p = o(d, this.view.state.doc);
		if (!f || !p || !f.posBeforeNode) throw Error("Column not found");
		return {
			type: "hover",
			leftColumn: {
				element: c,
				id: u,
				...f
			},
			rightColumn: {
				element: l,
				id: d,
				...p
			}
		};
	};
	mouseDownHandler = (e) => {
		let t = this.getColumnHoverOrDefaultState(e);
		if (t.type === "default") return;
		e.preventDefault();
		let n = e.clientX, r = t.leftColumn.element.getBoundingClientRect().width, i = t.rightColumn.element.getBoundingClientRect().width, a = t.leftColumn.node.attrs.width, o = t.rightColumn.node.attrs.width;
		t = {
			type: "resize",
			startPos: n,
			leftColumn: {
				...t.leftColumn,
				widthPx: r,
				widthPercent: a
			},
			rightColumn: {
				...t.rightColumn,
				widthPx: i,
				widthPercent: o
			}
		}, this.view.dispatch(this.view.state.tr.setMeta(U, t)), this.editor.getExtension(f)?.freezeMenu();
	};
	mouseMoveHandler = (e) => {
		let t = U.getState(this.view.state);
		if (!t) return;
		if (t.type !== "resize") {
			let n = this.getColumnHoverOrDefaultState(e), r = t.type === "default" && n.type === "default", i = t.type !== "default" && n.type !== "default" && t.leftColumn.id === n.leftColumn.id && t.rightColumn.id === n.rightColumn.id;
			if (r || i || n.type === "hover" && this.editor.getExtension(f)?.store.state?.show) return;
			this.view.dispatch(this.view.state.tr.setMeta(U, n));
			return;
		}
		let n = (e.clientX - t.startPos) * t.leftColumn.widthPercent, r = (t.leftColumn.widthPx + n) / t.leftColumn.widthPx - 1, i = t.leftColumn.widthPercent + r, a = t.rightColumn.widthPercent - r;
		i < this.COLUMN_MIN_WIDTH_PERCENT ? (a -= this.COLUMN_MIN_WIDTH_PERCENT - i, i = this.COLUMN_MIN_WIDTH_PERCENT) : a < this.COLUMN_MIN_WIDTH_PERCENT && (i -= this.COLUMN_MIN_WIDTH_PERCENT - a, a = this.COLUMN_MIN_WIDTH_PERCENT), this.view.dispatch(this.view.state.tr.setNodeAttribute(t.leftColumn.posBeforeNode, "width", i).setNodeAttribute(t.rightColumn.posBeforeNode, "width", a).setMeta("addToHistory", !1));
	};
	mouseUpHandler = (e) => {
		let t = U.getState(this.view.state);
		if (!t || t.type !== "resize") return;
		let n = this.getColumnHoverOrDefaultState(e);
		this.view.dispatch(this.view.state.tr.setMeta(U, n)), this.editor.getExtension(f)?.unfreezeMenu();
	};
	destroy() {
		this.view.dom.removeEventListener("mousedown", this.mouseDownHandler), document.body.removeEventListener("mousemove", this.mouseMoveHandler), document.body.removeEventListener("mouseup", this.mouseUpHandler);
	}
}, G = (e) => new c({
	key: U,
	props: { decorations: (e) => {
		let t = U.getState(e);
		return !t || t.type === "default" ? h.empty : h.create(e.doc, [m.node(t.leftColumn.posBeforeNode, t.leftColumn.posBeforeNode + t.leftColumn.node.nodeSize, { style: "box-shadow: 4px 0 0 #ccc; cursor: col-resize" }), m.node(t.rightColumn.posBeforeNode, t.rightColumn.posBeforeNode + t.rightColumn.node.nodeSize, { style: "cursor: col-resize" })]);
	} },
	state: {
		init: () => ({ type: "default" }),
		apply: (e, t) => {
			let n = e.getMeta(U);
			return n === void 0 ? t : n;
		}
	},
	view: (t) => new W(e, t)
}), K = (e) => u.create({
	name: "columnResize",
	addProseMirrorPlugins() {
		return [G(e)];
	}
}), q = d.create({
	name: "column",
	group: "bnBlock childContainer",
	content: "blockContainer+",
	priority: 40,
	defining: !0,
	marks: "deletion insertion modification",
	addAttributes() {
		return { width: {
			default: 1,
			parseHTML: (e) => {
				let t = e.getAttribute("data-width");
				if (t === null) return null;
				let n = parseFloat(t);
				return isFinite(n) ? n : null;
			},
			renderHTML: (e) => ({
				"data-width": e.width.toString(),
				style: `flex-grow: ${e.width};`
			})
		} };
	},
	parseHTML() {
		return [{
			tag: "div",
			getAttrs: (e) => typeof e == "string" ? !1 : e.getAttribute("data-node-type") === this.name ? {} : !1
		}];
	},
	renderHTML({ HTMLAttributes: e }) {
		let t = document.createElement("div");
		t.className = "bn-block-column", t.setAttribute("data-node-type", this.name);
		for (let [n, r] of Object.entries(e)) t.setAttribute(n, r);
		return {
			dom: t,
			contentDOM: t
		};
	},
	addExtensions() {
		return [K(this.options.editor)];
	}
}), J = d.create({
	name: "columnList",
	group: "childContainer bnBlock blockGroupChild",
	content: "column column+",
	priority: 40,
	defining: !0,
	marks: "deletion insertion modification",
	parseHTML() {
		return [{
			tag: "div",
			getAttrs: (e) => typeof e == "string" ? !1 : e.getAttribute("data-node-type") === this.name ? {} : !1
		}];
	},
	renderHTML({ HTMLAttributes: e }) {
		let t = document.createElement("div");
		t.className = "bn-block-column-list", t.setAttribute("data-node-type", this.name);
		for (let [n, r] of Object.entries(e)) t.setAttribute(n, r);
		return t.style.display = "flex", {
			dom: t,
			contentDOM: t
		};
	}
}), Y = n({
	node: q,
	type: "column",
	content: "none"
}, { width: { default: 1 } }, [H()]), X = n({
	node: J,
	type: "columnList",
	content: "none"
}, {}), Z = e.create({ blockSpecs: {
	column: Y,
	columnList: X
} }), Q = (e) => e.extend({ blockSpecs: {
	column: Y,
	columnList: X
} });
//#endregion
//#region src/extensions/SuggestionMenu/getMultiColumnSlashMenuItems.tsx
function $(e) {
	return "column" in e.schema.blockSchema && e.schema.blockSchema.columnList === Z.blockSchema.columnList && "column" in e.schema.blockSchema && e.schema.blockSchema.column === Z.blockSchema.column;
}
function ee(e) {
	let t = [];
	return $(e) && t.push({
		...L(e).slash_menu.two_columns,
		icon: /* @__PURE__ */ v(g, { size: 18 }),
		onItemClick: () => {
			p(e, {
				type: "columnList",
				children: [{
					type: "column",
					children: [{ type: "paragraph" }]
				}, {
					type: "column",
					children: [{ type: "paragraph" }]
				}]
			});
		}
	}, {
		...L(e).slash_menu.three_columns,
		icon: /* @__PURE__ */ v(_, { size: 18 }),
		onItemClick: () => {
			p(e, {
				type: "columnList",
				children: [
					{
						type: "column",
						children: [{ type: "paragraph" }]
					},
					{
						type: "column",
						children: [{ type: "paragraph" }]
					},
					{
						type: "column",
						children: [{ type: "paragraph" }]
					}
				]
			});
		}
	}), t;
}
//#endregion
export { Y as ColumnBlock, X as ColumnListBlock, $ as checkMultiColumnBlocksInSchema, z as detectEdgePosition, L as getMultiColumnDictionary, ee as getMultiColumnSlashMenuItems, I as locales, B as multiColumnDropCursor, Z as multiColumnSchema, Q as withMultiColumn };
