// @blocknote/xl-pdf-exporter/dist/blocknote-xl-pdf-exporter.js
import { Document as e, Font as t, Image as n, Link as r, Page as i, Path as a, StyleSheet as o, Svg as s, Text as c, View as l } from "@react-pdf/renderer";
import { Fragment as u, jsx as d, jsxs as f } from "react/jsx-runtime";
import { COLORS_DEFAULT as p, Exporter as m, mapTableCell as h } from "@blocknote/core";
import { Fragment as g } from "react";
//#region src/pdf/util/listItem.tsx
var _ = o.create({
	listItem: {
		display: "flex",
		flexDirection: "row",
		gap: 8 * .75,
		paddingRight: 10
	},
	bullet: {}
}), v = /* @__PURE__ */ d(s, {
	style: { marginTop: 2 },
	height: "12",
	viewBox: "0 -960 960 960",
	width: "12",
	fill: "undefined",
	children: /* @__PURE__ */ d(a, { d: "M472-480 332-620q-18-18-18-44t18-44q18-18 44-18t44 18l183 183q9 9 14 21t5 24q0 12-5 24t-14 21L420-252q-18 18-44 18t-44-18q-18-18-18-44t18-44l140-140Z" })
}), y = /* @__PURE__ */ d(s, {
	style: { marginTop: 2 },
	height: "12",
	viewBox: "0 -960 960 960",
	width: "12",
	fill: "undefined",
	children: /* @__PURE__ */ d(a, { d: "M200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200Zm0-80h560v-560H200v560Z" })
}), b = /* @__PURE__ */ d(s, {
	style: { marginTop: 2 },
	height: "12",
	viewBox: "0 -960 960 960",
	width: "12",
	fill: "undefined",
	children: /* @__PURE__ */ d(a, { d: "m424-312 282-282-56-56-226 226-114-114-56 56 170 170ZM200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200Zm0-80h560v-560H200v560Zm0-560v560-560Z" })
}), x = ({ listMarker: e, children: t, style: n }) => /* @__PURE__ */ f(l, {
	style: [_.listItem, n || {}],
	children: [/* @__PURE__ */ d(l, {
		style: _.bullet,
		children: typeof e == "string" ? /* @__PURE__ */ d(c, { children: e }) : e
	}), t]
}), S = .75, C = o.create({
	tableContainer: {},
	row: {
		flexDirection: "row",
		flexWrap: "wrap",
		display: "flex"
	},
	cell: {
		paddingHorizontal: 5 * S,
		paddingTop: 3 * S,
		borderLeft: "1px solid #ddd",
		borderTop: "1px solid #ddd",
		wordWrap: "break-word",
		whiteSpace: "pre-wrap"
	},
	headerCell: { fontWeight: "bold" },
	bottomCell: { borderBottom: "1px solid #ddd" },
	rightCell: { borderRight: "1px solid #ddd" }
}), w = (e) => {
	let t = Array(e.data.headerRows ?? 0).fill(!0), n = Array(e.data.headerCols ?? 0).fill(!0);
	return /* @__PURE__ */ d(l, {
		style: C.tableContainer,
		wrap: !1,
		children: e.data.rows.map((r, i) => /* @__PURE__ */ d(l, {
			style: [C.row, i === e.data.rows.length - 1 ? C.bottomCell : {}],
			children: r.cells.map((a, o) => {
				let s = h(a), c = t[i], u = n[o];
				return /* @__PURE__ */ d(l, {
					style: [
						C.cell,
						c || u ? C.headerCell : {},
						o === r.cells.length - 1 ? C.rightCell : {},
						e.data.columnWidths[o] ? { width: e.data.columnWidths[o] } : { flex: 1 },
						{
							color: s.props.textColor === "default" ? void 0 : e.transformer.options.colors[s.props.textColor]?.text,
							backgroundColor: s.props.backgroundColor === "default" ? void 0 : e.transformer.options.colors[s.props.backgroundColor]?.background,
							textAlign: s.props.textAlignment
						}
					],
					children: e.transformer.transformInlineContent(s.content)
				}, o);
			})
		}, i))
	});
}, T = .75, E = 16, D = {
	paragraph: (e, t) => /* @__PURE__ */ d(c, { children: t.transformInlineContent(e.content) }, "paragraph" + e.id),
	toggleListItem: (e, t) => /* @__PURE__ */ d(x, {
		listMarker: v,
		children: /* @__PURE__ */ d(c, { children: t.transformInlineContent(e.content) })
	}),
	bulletListItem: (e, t) => /* @__PURE__ */ d(x, {
		listMarker: "•",
		children: /* @__PURE__ */ d(c, { children: t.transformInlineContent(e.content) })
	}, "bulletListItem" + e.id),
	numberedListItem: (e, t, n, r) => /* @__PURE__ */ d(x, {
		listMarker: `${r}.`,
		children: /* @__PURE__ */ d(c, { children: t.transformInlineContent(e.content) })
	}, "numberedListItem" + e.id),
	checkListItem: (e, t) => /* @__PURE__ */ d(x, {
		listMarker: e.props.checked ? b : y,
		children: /* @__PURE__ */ d(c, { children: t.transformInlineContent(e.content) })
	}, "checkListItem" + e.id),
	heading: (e, t) => {
		let n = {
			1: 2,
			2: 1.5,
			3: 1.17,
			4: 1,
			5: .83,
			6: .67
		}[e.props.level];
		return /* @__PURE__ */ d(c, {
			style: {
				fontSize: n * E * T,
				lineHeight: 1.25,
				fontWeight: 700
			},
			children: t.transformInlineContent(e.content)
		}, "heading" + e.id);
	},
	quote: (e, t) => /* @__PURE__ */ d(c, {
		style: {
			borderLeft: "#7D797A",
			color: "#7D797A",
			paddingLeft: 9.5 * T
		},
		children: t.transformInlineContent(e.content)
	}, "quote" + e.id),
	codeBlock: (e) => {
		let t = e.content.map((e) => e.text).join("").split("\n").map((t, n) => /* @__PURE__ */ d(c, {
			style: { marginLeft: (t.match(/^\s*/)?.[0].length || 0) * 9.5 * T },
			children: t.trimStart() || /* @__PURE__ */ d(u, { children: "\xA0" })
		}, `line_${n}` + e.id));
		return /* @__PURE__ */ d(l, {
			wrap: !1,
			style: {
				padding: 24 * T,
				border: "1px solid #000000",
				lineHeight: 1.25,
				fontSize: E * T,
				fontFamily: "GeistMono"
			},
			children: t
		}, "codeBlock" + e.id);
	},
	pageBreak: () => /* @__PURE__ */ d(l, { break: !0 }, "pageBreak"),
	divider: () => /* @__PURE__ */ d(l, { style: {
		borderTop: "1px solid #ccc",
		marginTop: 11.5 * T,
		marginBottom: 11.5 * T
	} }),
	column: (e, t, n, r, i) => /* @__PURE__ */ d(l, {
		style: { flex: e.props.width },
		children: i
	}),
	columnList: (e, t, n, r, i) => /* @__PURE__ */ d(l, {
		style: {
			display: "flex",
			gap: 8 * T,
			flexDirection: "row"
		},
		children: i
	}),
	audio: (e, t) => /* @__PURE__ */ f(l, {
		wrap: !1,
		children: [O(e.props, "Open audio file", /* @__PURE__ */ d(s, {
			height: 14,
			width: 14,
			viewBox: "0 0 24 24",
			fill: "currentColor",
			children: /* @__PURE__ */ d(a, { d: "M2 16.0001H5.88889L11.1834 20.3319C11.2727 20.405 11.3846 20.4449 11.5 20.4449C11.7761 20.4449 12 20.2211 12 19.9449V4.05519C12 3.93977 11.9601 3.8279 11.887 3.73857C11.7121 3.52485 11.3971 3.49335 11.1834 3.66821L5.88889 8.00007H2C1.44772 8.00007 1 8.44778 1 9.00007V15.0001C1 15.5524 1.44772 16.0001 2 16.0001ZM23 12C23 15.292 21.5539 18.2463 19.2622 20.2622L17.8445 18.8444C19.7758 17.1937 21 14.7398 21 12C21 9.26016 19.7758 6.80629 17.8445 5.15557L19.2622 3.73779C21.5539 5.75368 23 8.70795 23 12ZM18 12C18 10.0883 17.106 8.38548 15.7133 7.28673L14.2842 8.71584C15.3213 9.43855 16 10.64 16 12C16 13.36 15.3213 14.5614 14.2842 15.2841L15.7133 16.7132C17.106 15.6145 18 13.9116 18 12Z" })
		}), t), k(e.props, t)]
	}, "audio" + e.id),
	video: (e, t) => /* @__PURE__ */ f(l, {
		wrap: !1,
		children: [O(e.props, "Open video file", /* @__PURE__ */ d(s, {
			height: 14,
			width: 14,
			viewBox: "0 0 24 24",
			fill: "currentColor",
			children: /* @__PURE__ */ d(a, { d: "M2 3.9934C2 3.44476 2.45531 3 2.9918 3H21.0082C21.556 3 22 3.44495 22 3.9934V20.0066C22 20.5552 21.5447 21 21.0082 21H2.9918C2.44405 21 2 20.5551 2 20.0066V3.9934ZM8 5V19H16V5H8ZM4 5V7H6V5H4ZM18 5V7H20V5H18ZM4 9V11H6V9H4ZM18 9V11H20V9H18ZM4 13V15H6V13H4ZM18 13V15H20V13H18ZM4 17V19H6V17H4ZM18 17V19H20V17H18Z" })
		}), t), k(e.props, t)]
	}, "video" + e.id),
	file: (e, t) => /* @__PURE__ */ f(l, {
		wrap: !1,
		children: [O(e.props, "Open file", /* @__PURE__ */ d(s, {
			height: 16,
			width: 16,
			viewBox: "0 0 24 24",
			fill: "currentColor",
			children: /* @__PURE__ */ d(a, { d: "M3 8L9.00319 2H19.9978C20.5513 2 21 2.45531 21 2.9918V21.0082C21 21.556 20.5551 22 20.0066 22H3.9934C3.44476 22 3 21.5501 3 20.9932V8ZM10 4V9H5V20H19V4H10Z" })
		}), t), k(e.props, t)]
	}, "file" + e.id),
	image: async (e, t) => /* @__PURE__ */ f(l, {
		wrap: !1,
		children: [/* @__PURE__ */ d(n, {
			src: await t.resolveFile(e.props.url),
			style: { width: e.props.previewWidth ? e.props.previewWidth * T : void 0 }
		}), k(e.props, t)]
	}, "image" + e.id),
	table: (e, t) => /* @__PURE__ */ d(w, {
		data: e.content,
		transformer: t
	}, "table" + e.id)
};
function O(e, t, n, i) {
	return /* @__PURE__ */ d(r, {
		src: e.url,
		children: /* @__PURE__ */ f(l, {
			style: {
				display: "flex",
				gap: 8 * .75,
				flexDirection: "row"
			},
			children: [n, /* @__PURE__ */ d(c, { children: e.name || t })]
		})
	}, "file" + e.url);
}
function k(e, t) {
	if (e.caption) return /* @__PURE__ */ d(c, {
		style: {
			width: e.previewWidth ? e.previewWidth * T : void 0,
			fontSize: E * .8 * T
		},
		children: e.caption
	}, "caption" + e.caption);
}
//#endregion
//#region src/pdf/defaultSchema/index.ts
var A = {
	blockMapping: D,
	inlineContentMapping: {
		link: (e, t) => /* @__PURE__ */ d(r, {
			href: e.href,
			children: e.content.map((e) => t.transformStyledText(e))
		}, "link" + e.href),
		text: (e, t) => t.transformStyledText(e)
	},
	styleMapping: {
		bold: (e) => e ? { fontWeight: "bold" } : {},
		italic: (e) => e ? { fontStyle: "italic" } : {},
		underline: (e) => e ? { textDecoration: "underline" } : {},
		strike: (e) => e ? { textDecoration: "line-through" } : {},
		backgroundColor: (e, t) => e ? { backgroundColor: t.options.colors[e]?.background } : {},
		textColor: (e, t) => e ? { color: t.options.colors[e]?.text } : {},
		code: (e) => e ? { fontFamily: "GeistMono" } : {}
	}
};
//#endregion
//#region ../../shared/api/corsProxy.ts
async function j(e) {
	return "https://corsproxy.api.blocknotejs.org/corsproxy/?url=" + encodeURIComponent(e);
}
//#endregion
//#region ../../shared/util/fileUtil.ts
async function M(e, t) {
	return e.default;
}
async function N(e) {
	return M(e, "font/ttf");
}
//#endregion
//#region src/pdf/pdfExporter.tsx
var P = 16, F = .75, I = class extends m {
	fontsRegistered = !1;
	styles = o.create({
		page: {
			paddingTop: 35,
			paddingBottom: 65,
			paddingHorizontal: 35,
			fontFamily: "Inter",
			fontSize: P * F,
			lineHeight: 1.5
		},
		block: {},
		blockChildren: {},
		header: {},
		footer: { position: "absolute" }
	});
	options;
	constructor(e, t, n) {
		let r = {
			emojiSource: {
				format: "png",
				url: "https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/72x72/"
			},
			resolveFileUrl: j,
			colors: p,
			...n
		};
		super(e, t, r), this.schema = e, this.options = r;
	}
	transformStyledText(e) {
		let t = this.mapStyles(e.styles);
		return /* @__PURE__ */ d(c, {
			style: Object.assign({}, ...t),
			children: e.text
		}, e.text);
	}
	async transformBlocks(e, t = 0) {
		let n = [], r = 0;
		for (let i of e) {
			i.type === "numberedListItem" ? r++ : r = 0;
			let e = await this.transformBlocks(i.children, t + 1), a = await this.mapBlock(i, t, r, e);
			if ([
				"pageBreak",
				"columnList",
				"column"
			].includes(i.type)) {
				n.push(a);
				continue;
			}
			let o = this.blocknoteDefaultPropsToReactPDFStyle(i.props);
			n.push(/* @__PURE__ */ f(g, { children: [/* @__PURE__ */ d(l, {
				style: {
					paddingVertical: 3 * F,
					...this.styles.block,
					...o
				},
				children: a
			}), e.length > 0 && /* @__PURE__ */ d(l, {
				style: {
					marginLeft: P * 1.5 * F,
					...this.styles.blockChildren
				},
				children: e
			}, i.id + t + "children")] }, i.id));
		}
		return n;
	}
	async registerFonts() {
		if (this.fontsRegistered) return;
		this.options.emojiSource && t.registerEmojiSource(this.options.emojiSource);
		let e = await N(await import("./Inter_18pt-Regular-DjtEslHp.js"));
		t.register({
			family: "Inter",
			src: e
		}), e = await N(await import("./Inter_18pt-Italic-B09-NoiJ.js")), t.register({
			family: "Inter",
			fontStyle: "italic",
			src: e
		}), e = await N(await import("./Inter_18pt-Bold-CAT0xCdm.js")), t.register({
			family: "Inter",
			src: e,
			fontWeight: "bold"
		}), e = await N(await import("./Inter_18pt-BoldItalic-CUvI6MuN.js")), t.register({
			family: "Inter",
			fontStyle: "italic",
			src: e,
			fontWeight: "bold"
		}), e = await N(await import("./GeistMono-Regular-D2zmhEYE.js")), t.register({
			family: "GeistMono",
			src: e
		}), this.fontsRegistered = !0;
	}
	async toReactPDFDocument(t, n = {}) {
		return await this.registerFonts(), /* @__PURE__ */ d(e, { children: /* @__PURE__ */ f(i, {
			dpi: 100,
			size: "A4",
			style: this.styles.page,
			children: [
				n.header && /* @__PURE__ */ d(l, {
					fixed: !0,
					style: this.styles.header,
					children: n.header
				}),
				await this.transformBlocks(t),
				n.footer && /* @__PURE__ */ d(l, {
					fixed: !0,
					style: [{
						left: this.styles.page.paddingHorizontal || 0,
						bottom: (this.styles.page.paddingBottom || 0) / 2,
						right: this.styles.page.paddingHorizontal || 0
					}, this.styles.footer],
					children: n.footer
				})
			]
		}) });
	}
	blocknoteDefaultPropsToReactPDFStyle(e) {
		return {
			textAlign: e.textAlignment,
			backgroundColor: e.backgroundColor === "default" || !e.backgroundColor ? void 0 : this.options.colors[e.backgroundColor]?.background,
			color: e.textColor === "default" || !e.textColor ? void 0 : this.options.colors[e.textColor]?.text,
			alignItems: e.textAlignment === "right" ? "flex-end" : e.textAlignment === "center" ? "center" : void 0
		};
	}
};
//#endregion
export { I as PDFExporter, A as pdfDefaultSchemaMappings };
